File: dev/requirements/shared/RequirementCase.swift
import Foundation

/// One executable proof of one requirement leaf. The id is the spec's dotted
/// number; the coverage gate is what guarantees it exists and is claimed once.
struct RequirementCase {
    let id: String
    let verify: () throws -> Void

    init(id: String, verify: @escaping () throws -> Void) {
        self.id = id
        self.verify = verify
    }
}

struct RequirementFailure: Error, CustomStringConvertible {
    let description: String
}

func expect(_ condition: Bool, _ message: @autoclosure () -> String) throws {
    if !condition { throw RequirementFailure(description: message()) }
}

func expectEqual<T: Equatable>(_ actual: T, _ expected: T, _ label: String) throws {
    if actual != expected {
        throw RequirementFailure(description: "\(label): expected \(expected), got \(actual)")
    }
}

func expectNil<T>(_ actual: T?, _ label: String) throws {
    if let actual { throw RequirementFailure(description: "\(label): expected nothing, got \(actual)") }
}

File: dev/requirements/shared/requirement_case.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';

/// One executable proof of one requirement leaf.
///
/// The id is the leaf it claims, and the coverage gate is what keeps that
/// honest: a case claiming an id no requirement carries, or a second case
/// claiming an id another already has, fails the build rather than passing
/// quietly.
class RequirementCase {
  const RequirementCase(this.id, this.verify);

  /// The dotted leaf id from `requirements.md`.
  final String id;

  /// Every kind takes the tester, whether or not it renders anything: it is
  /// what lets one runner execute all of them, and a rule that needs no
  /// renderer simply ignores it.
  final FutureOr<void> Function(WidgetTester tester) verify;
}
